### PR TITLE
Support basic cases in UseRangeOperator

### LINKED ISSUE #45637: "Substring can be simplified" misses basic cases
**Version Used**: 
IDE0057 handles morecomplicated cases of Substring, but not the basic cases.

**Steps to Reproduce**:

``` C#
var value = "Hello World"
value = value.Substring(1);
```

**Expected Behavior**:
A analyzer message, asking me to change above to:

``` C#
var value = "Hello World"
value = value[1...];
```

**Actual Behavior**:
Nothing

### PR DESCRIPTION
**Pull request by mahalex**
Fixes #45637

### CODE REVIEW

File: src/Analyzers/CSharp/Analyzers/UseIndexOrRangeOperator/CSharpUseRangeOperatorDiagnosticAnalyzer.InfoCache.cs [hunk 1]
```diff
@@ -69,6 +69,42 @@ public bool TryGetMemberInfo(IMethodSymbol method, out MemberInfo memberInfo)
...
 
+            public bool TryGetMemberInfoOneArgument(IMethodSymbol method, out MemberInfo memberInfo)
+            {
+                if (!IsSliceLikeMethodWithOneArgument(method))
+                {
+                    memberInfo = default;
+                    return false;
+                }
+
+                if (_methodToMemberInfo.TryGetValue(method, out memberInfo))
```
Comment (CyrusNajmabadi) [inline src/Analyzers/CSharp/Analyzers/UseIndexOrRangeOperator/CSharpUseRangeOperatorDiagnosticAnalyzer.InfoCache.cs:80]:
```suggestion
                if (!_methodToMemberInfo.TryGetValue(method, out memberInfo))
```

then have hte logic to try to produce the value in the if-statement.

File: src/Analyzers/CSharp/Analyzers/UseIndexOrRangeOperator/CSharpUseRangeOperatorDiagnosticAnalyzer.Result.cs [hunk 2]
```diff
@@ -28,6 +28,10 @@ public enum ResultKind
             public readonly IMethodSymbol SliceLikeMethod;
             public readonly MemberInfo MemberInfo;
             public readonly IOperation Op1;
+
+            /// <summary>
+            /// Can be null, if we are dealing with one-argument call to a slice-like method.
+            /// </summary>
             public readonly IOperation Op2;
```
Comment (CyrusNajmabadi) [inline src/Analyzers/CSharp/Analyzers/UseIndexOrRangeOperator/CSharpUseRangeOperatorDiagnosticAnalyzer.Result.cs:35]:
```suggestion
            public readonly IOperation? Op2;
```

File: src/Analyzers/CSharp/Analyzers/UseIndexOrRangeOperator/CSharpUseRangeOperatorDiagnosticAnalyzer.cs [hunk 3]
```diff
@@ -111,13 +111,54 @@ protected override void InitializeWorker(AnalysisContext context)
...
+            IInvocationOperation invocation,
+            InfoCache infoCache,
+            InvocationExpressionSyntax invocationSyntax,
+            CodeStyleOption2<bool> option)
+        {
+            // See if the call is to something slice-like.
+            var targetMethod = invocation.TargetMethod;
+
+            // Try to see if we're calling into some sort of Slice method with a matching
+            // indexer or overload
```
Comment (CyrusNajmabadi) [inline src/Analyzers/CSharp/Analyzers/UseIndexOrRangeOperator/CSharpUseRangeOperatorDiagnosticAnalyzer.cs:138]: i'm not finding tehse comments helpful.  having real world examples of the before/after would be more helpful here imo.

File: src/Analyzers/CSharp/Analyzers/UseIndexOrRangeOperator/CSharpUseRangeOperatorDiagnosticAnalyzer.cs [hunk 4]
```diff
@@ -111,13 +111,54 @@ protected override void InitializeWorker(AnalysisContext context)
...
+            var startOperation = invocation.Arguments[0].Value;
+            return new Result(
+                ResultKind.Computed,
+                option,
+                invocation,
+                invocationSyntax,
+                targetMethod,
+                memberInfo,
+                startOperation,
+                null); // secondOperation is null: the range will run to the end.
```
Comment (CyrusNajmabadi) [inline src/Analyzers/CSharp/Analyzers/UseIndexOrRangeOperator/CSharpUseRangeOperatorDiagnosticAnalyzer.cs:153]: named param.

File: src/Analyzers/CSharp/Analyzers/UseIndexOrRangeOperator/CSharpUseRangeOperatorDiagnosticAnalyzer.cs [hunk 5]
```diff
@@ -111,13 +111,54 @@ protected override void InitializeWorker(AnalysisContext context)
...
+        }
+
+        private static Result? AnalyzeInvocationWithTwoArguments(
```
Comment (CyrusNajmabadi) [inline src/Analyzers/CSharp/Analyzers/UseIndexOrRangeOperator/CSharpUseRangeOperatorDiagnosticAnalyzer.cs:156]:
```suggestion
        private static Result? AnalyzeTwoArgumentInvocation(
```

File: src/Analyzers/CSharp/Analyzers/UseIndexOrRangeOperator/Helpers.cs [hunk 6]
```diff
@@ -91,6 +91,21 @@ public static bool IsSliceLikeMethod(IMethodSymbol method)
...
+        /// This name was picked by examining the patterns in the BCL for slicing members.
+        /// </summary>
+        public static bool IsSliceLikeMethodWithOneArgument(IMethodSymbol method)
+            => method != null &&
+               IsPublicInstance(method) &&
+               method.Parameters.Length == 1 &&
+               // From: https://github.com/dotnet/csharplang/blob/master/proposals/csharp-8.0/ranges.md#decisions-made-during-implementation
+               //
+               // When looking for the pattern members, we look for original definitions, not
+               // constructed members
```
Comment (CyrusNajmabadi) [inline src/Analyzers/CSharp/Analyzers/UseIndexOrRangeOperator/Helpers.cs:106]: just make this a normal method.  breaking up an expression with complex comments like this just makes things too hard to read.

File: src/Analyzers/CSharp/CodeFixes/UseIndexOrRangeOperator/CSharpUseRangeOperatorCodeFixProvider.cs [hunk 7]
```diff
@@ -150,15 +150,28 @@ private static RangeExpressionSyntax CreateComputedRange(Result result)
...
-            // Similarly, if our end-op is actually equivalent to `expr.Length - val`, then just
-            // change our end-op to be `val` and record that we should emit it as `^val`.
-            var endFromEnd = IsFromEnd(lengthLikeProperty, instance, ref endOperation);
-            var endExpr = (ExpressionSyntax)endOperation.Syntax;
-
-            // If the range operation goes to 'expr.Length' then we can just leave off the end part
-            // of the range.  i.e. `start..`
-            if (IsInstanceLengthCheck(lengthLikeProperty, instance, endOperation))
+            var endFromEnd = false;
+            ExpressionSyntax endExpr;
```
Comment (CyrusNajmabadi) [inline src/Analyzers/CSharp/CodeFixes/UseIndexOrRangeOperator/CSharpUseRangeOperatorCodeFixProvider.cs:154]: can we just make this null, and then just set it in the case where we have it?

File: src/Analyzers/CSharp/Analyzers/UseIndexOrRangeOperator/CSharpUseRangeOperatorDiagnosticAnalyzer.InfoCache.cs [hunk 8]
```diff
@@ -69,6 +69,42 @@ public bool TryGetMemberInfo(IMethodSymbol method, out MemberInfo memberInfo)
...
+                }
+
+                // Find overload of our method that is a slice-like method with two arguments.
+                // Computing member info for this method will also check that the containing type
+                // has an int32 'Length' or 'Count' property, and has a suitable indexer,
+                // so we don't have to.
+                var overloadWithTwoArguments = method.ContainingType
+                    .GetMembers(method.Name)
+                    .OfType<IMethodSymbol>()
+                    .FirstOrDefault(s => IsSliceLikeMethod(s));
```
Comment (CyrusNajmabadi) [inline src/Analyzers/CSharp/Analyzers/UseIndexOrRangeOperator/CSharpUseRangeOperatorDiagnosticAnalyzer.InfoCache.cs:92]: should we rename IsSliceLikeMethod to IsTwoArgumentSliceLikeMethod?

**Comment by CyrusNajmabadi** [comment]
Nice start!

File: src/Analyzers/CSharp/Analyzers/UseIndexOrRangeOperator/CSharpUseRangeOperatorDiagnosticAnalyzer.InfoCache.cs (hunk 1 above)
Comment (mahalex) [reply src/Analyzers/CSharp/Analyzers/UseIndexOrRangeOperator/CSharpUseRangeOperatorDiagnosticAnalyzer.InfoCache.cs]: Fixed.

File: src/Analyzers/CSharp/Analyzers/UseIndexOrRangeOperator/CSharpUseRangeOperatorDiagnosticAnalyzer.Result.cs (hunk 2 above)
Comment (mahalex) [reply src/Analyzers/CSharp/Analyzers/UseIndexOrRangeOperator/CSharpUseRangeOperatorDiagnosticAnalyzer.Result.cs]: Fixed.

File: src/Analyzers/CSharp/Analyzers/UseIndexOrRangeOperator/CSharpUseRangeOperatorDiagnosticAnalyzer.cs (hunk 3 above)
Comment (mahalex) [reply src/Analyzers/CSharp/Analyzers/UseIndexOrRangeOperator/CSharpUseRangeOperatorDiagnosticAnalyzer.cs]: Agreed. I re-worded the comment.

File: src/Analyzers/CSharp/Analyzers/UseIndexOrRangeOperator/CSharpUseRangeOperatorDiagnosticAnalyzer.cs (hunk 4 above)
Comment (mahalex) [reply src/Analyzers/CSharp/Analyzers/UseIndexOrRangeOperator/CSharpUseRangeOperatorDiagnosticAnalyzer.cs]: Fixed.

File: src/Analyzers/CSharp/Analyzers/UseIndexOrRangeOperator/CSharpUseRangeOperatorDiagnosticAnalyzer.InfoCache.cs (hunk 8 above)
Comment (mahalex) [reply src/Analyzers/CSharp/Analyzers/UseIndexOrRangeOperator/CSharpUseRangeOperatorDiagnosticAnalyzer.InfoCache.cs]: Fixed; also renamed `IsSliceLikeMethodWithOneArgument` to `IsOneArgumentSliceLikeMethod`.

File: src/Analyzers/CSharp/CodeFixes/UseIndexOrRangeOperator/CSharpUseRangeOperatorCodeFixProvider.cs (hunk 7 above)
Comment (mahalex) [reply src/Analyzers/CSharp/CodeFixes/UseIndexOrRangeOperator/CSharpUseRangeOperatorCodeFixProvider.cs]: Fixed.

File: src/Analyzers/CSharp/Analyzers/UseIndexOrRangeOperator/CSharpUseRangeOperatorDiagnosticAnalyzer.cs (hunk 5 above)
Comment (mahalex) [reply src/Analyzers/CSharp/Analyzers/UseIndexOrRangeOperator/CSharpUseRangeOperatorDiagnosticAnalyzer.cs]: Fixed.

Comment (mahalex) [reply src/Analyzers/CSharp/Analyzers/UseIndexOrRangeOperator/CSharpUseRangeOperatorDiagnosticAnalyzer.cs]: Also renamed `AnalyzeInvocationWithOneArgument` to `AnalyzeOneArgumentInvocation`.

File: src/Analyzers/CSharp/Analyzers/UseIndexOrRangeOperator/Helpers.cs (hunk 6 above)
Comment (mahalex) [reply src/Analyzers/CSharp/Analyzers/UseIndexOrRangeOperator/Helpers.cs]: Fixed.

**Comment by CyrusNajmabadi** [comment]
Feel free to ping me when this passes, and i'll merge in.  Thanks!

**Comment by mahalex** [comment]
@CyrusNajmabadi thanks for the feedback! It’s ready for merging now.

**Comment by CyrusNajmabadi** [comment]
Thank you for the contribution!

**Comment by davkean** [comment]
@mahalex Nice!

**Pull request merged**